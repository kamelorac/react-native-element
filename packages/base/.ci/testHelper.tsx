import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
