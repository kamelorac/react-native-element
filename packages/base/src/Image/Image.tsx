import React, { useCallback } from 'react';
import {
  Animated,
  Image as ImageNative,
  ImageLoadEventData,
  ImageProps as RNImageProps,
  NativeSyntheticEvent,
  StyleProp,
  StyleSheet,
  View,
  Pressable,
  ViewStyle,
  Text,
} from 'react-native';
import { InlinePressableProps, RneFunctionComponent } from '../helpers';

export interface ImageProps extends RNImageProps, InlinePressableProps {
  /** Define the component passed to image.
   *  @default `Press handlers present then Pressable else View`
   */
  Component?: typeof React.Component;

  /** Specify a different component as the Image component. */
  ImageComponent?: typeof React.Component;

  /** Content to load when Image is rendering.
   */
  PlaceholderContent?: React.ReactElement;

  /** Additional styling for the container. */
  containerStyle?: StyleProp<ViewStyle>;

  /** Additional styling for the children container. */
  childrenContainerStyle?: StyleProp<ViewStyle>;

  /** Additional styling for the placeholder container. */
  placeholderStyle?: StyleProp<ViewStyle>;

  /** Perform fade transition on image load. */
  transition?: boolean;

  /** Perform fade transition on image load. */
  transitionDuration?: number;
}

/** Drop-in replacement for the standard React Native Image component that displays
images with a placeholder and smooth image load transitioning. */
export const Image: RneFunctionComponent<ImageProps> = ({
  onPress,
  onLongPress,
  onPressIn,
  onPressOut,
  Component = onPress || onLongPress || onPressIn || onPressOut
    ? Pressable
    : View,
  placeholderStyle,
  PlaceholderContent,
  containerStyle,
  childrenContainerStyle = null,
  style = {},
  ImageComponent = ImageNative,
  onLoad,
  children,
  transition,
  transitionDuration = 360,
  pressableProps,
  ...props
}) => {
  const placeholderOpacity = React.useRef(new Animated.Value(1));

  const onLoadHandler = useCallback(
    (event: NativeSyntheticEvent<ImageLoadEventData>) => {
      if (transition) {
        Animated.timing(placeholderOpacity.current, {
          toValue: 0,
          duration: transitionDuration,
          useNativeDriver: true,
        }).start();
      } else {
        placeholderOpacity.current.setValue(0);
      }
      onLoad?.(event);
    },
    [transition, transitionDuration, onLoad]
  );

  const hasImage = Boolean(props.source);

  return (
    <Component
      {...pressableProps}
      {...{ onPress, onPressIn, onPressOut, onLongPress }}
      accessibilityIgnoresInvertColors={true}
      style={StyleSheet.flatten([styles.container, containerStyle])}
    >
      <ImageComponent
        testID="RNE__Image"
        {...props}
        {...{ transition, transitionDuration }}
        onLoad={onLoadHandler}
        style={StyleSheet.flatten([StyleSheet.absoluteFill, style])}
      />
      {/* Transition placeholder */}
      <Animated.View
        pointerEvents={hasImage ? 'none' : 'auto'}
        accessibilityElementsHidden={hasImage}
        importantForAccessibility={hasImage ? 'no-hide-descendants' : 'yes'}
        style={[
          StyleSheet.absoluteFillObject,
          {
            opacity: hasImage ? placeholderOpacity.current : 1,
          },
        ]}
      >
        <View
          testID="RNE__Image__placeholder"
          style={StyleSheet.flatten([
            style,
            styles.placeholder,
            placeholderStyle,
          ])}
        >
          {React.isValidElement(PlaceholderContent)
            ? PlaceholderContent
            : PlaceholderContent && (
                <Text testID="RNE__Image__Placeholder__Content">
                  {PlaceholderContent}
                </Text>
              )}
        </View>
      </Animated.View>
      {/* Children for Image */}
      <View
        testID="RNE__Image__children__container"
        style={childrenContainerStyle ?? style}
      >
        {children}
      </View>
    </Component>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
    position: 'relative',
    overflow: 'hidden',
  },
  placeholder: {
    backgroundColor: '#bdbdbd',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

Image.displayName = 'Image';
