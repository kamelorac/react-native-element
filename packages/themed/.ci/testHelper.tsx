import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import {
  ThemeProvider,
  FullTheme,
  createTheme,
  CreateThemeOptions,
} from '../src/config';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<CreateThemeOptions> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    <ThemeProvider theme={createTheme(themeProp)}>{children}</ThemeProvider>,
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
