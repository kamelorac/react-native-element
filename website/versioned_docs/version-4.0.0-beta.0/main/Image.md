---
id: image

title: Image

slug: /image
---

import Usage from './usage/Image/Image.md'

Drop-in replacement for the standard React Native Image component that displays

images with a placeholder and smooth image load transitioning.

## Usage

<Usage />

---

## Props

### Image

- [Component](#component)
- [ImageComponent](#imagecomponent)
- [PlaceholderContent](#placeholdercontent)
- [childrenContainerStyle](#childrencontainerstyle)
- [containerStyle](#containerstyle)
- [onLongPress](#onlongpress)
- [onPress](#onpress)
- [placeholderStyle](#placeholderstyle)
- [pressableProps](#pressableprops)
- [transition](#transition)
- [transitionDuration](#transitionduration)

## Reference

### Image

#### Component

Define the component passed to image.

| Type            | Default |
| --------------- | ------- |
| React Component | None    |

---

#### ImageComponent

Specify a different component as the Image component.

| Type | Default |
| ---- | ------- |
| any  | None    |

---

#### PlaceholderContent

Content to load when Image is rendering.

| Type | Default |
| ---- | ------- |
| any  | None    |

---

#### childrenContainerStyle

Additional styling for the children container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | null    |

---

#### containerStyle

Additional styling for the container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### onLongPress

Called when a long-tap gesture is detected.

Callback function when long pressing component.

| Type                                                      | Default |
| --------------------------------------------------------- | ------- |
| ((event: GestureResponderEvent) => void) and (() => void) | None    |

---

#### onPress

Called when a single tap gesture is detected.

Callback function when pressing component.

| Type                                                      | Default |
| --------------------------------------------------------- | ------- |
| ((event: GestureResponderEvent) => void) and (() => void) | None    |

---

#### placeholderStyle

Additional styling for the placeholder container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### pressableProps

| Type                                                                            | Default |
| ------------------------------------------------------------------------------- | ------- |
| Omit<PressableProps, "onPress" or "onLongPress" or "onPressIn" or "onPressOut"> | None    |

---

#### transition

Perform fade transition on image load.

| Type    | Default |
| ------- | ------- |
| boolean | None    |

---

#### transitionDuration

Perform fade transition on image load.

| Type   | Default |
| ------ | ------- |
| number | 360     |

---
