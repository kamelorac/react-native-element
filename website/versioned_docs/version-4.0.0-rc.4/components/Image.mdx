---
id: image
title: Image
---

import { IconsStyle } from "@site/src/components/Docs_Icons";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import Usage from "../component_usage/Image.mdx";

<IconsStyle />

Drop-in replacement for the standard React Native Image component that displays
images with a placeholder and smooth image load transitioning.

## Import

```tsx
import { Image } from "@rneui/themed";
```

## Usage

<Usage />

## Props

:::note
Includes all [React Native Image](https://reactnative.dev/docs/image#props) props.
:::

<div class='table-responsive'>

| Name                     | Type                                                        | Default                                           | Description                                           |
| ------------------------ | ----------------------------------------------------------- | ------------------------------------------------- | ----------------------------------------------------- |
| `Component`              | React Component                                             | `Press handlers present then Pressable else View` | Define the component passed to image.                 |
| `ImageComponent`         | typeof Component                                            |                                                   | Specify a different component as the Image component. |
| `PlaceholderContent`     | `ReactElement<any, string` \| `JSXElementConstructor<any>>` |                                                   | Content to load when Image is rendering.              |
| `childrenContainerStyle` | View Style                                                  | `null`                                            | Additional styling for the children container.        |
| `containerStyle`         | View Style                                                  |                                                   | Additional styling for the container.                 |
| `onLongPress`            | GestureResponderEventHandler                                |                                                   | Called when a long-tap gesture is detected.           |
| `onPress`                | GestureResponderEventHandler                                |                                                   | Called when a single tap gesture is detected.         |
| `onPressIn`              | GestureResponderEventHandler                                |                                                   | Called when a touch is engaged before `onPress`.      |
| `onPressOut`             | GestureResponderEventHandler                                |                                                   | Called when a touch is released before `onPress`.     |
| `placeholderStyle`       | View Style                                                  |                                                   | Additional styling for the placeholder container.     |
| `pressableProps`         | PressableProps except click handlers                        | `None`                                            |                                                       |
| `transition`             | boolean                                                     |                                                   | Perform fade transition on image load.                |
| `transitionDuration`     | number                                                      | `360`                                             | Perform fade transition on image load.                |

</div>
