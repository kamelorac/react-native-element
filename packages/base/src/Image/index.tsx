import { Image as ImageNative } from 'react-native';
import { Image as RNEImage, ImageProps } from './Image';

const Image = Object.assign(RNEImage, {
  getSize: ImageNative.getSize,
  getSizeWithHeaders: ImageNative.getSizeWithHeaders,
  prefetch: ImageNative.prefetch,
  abortPrefetch: ImageNative.abortPrefetch,
  queryCache: ImageNative.queryCache,
  resolveAssetSource: ImageNative.resolveAssetSource,
});

export { Image };
export type { ImageProps };
