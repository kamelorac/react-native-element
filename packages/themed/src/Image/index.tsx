import { Image as ImageNative } from 'react-native';
import { withTheme } from '../config';
import { Image, ImageProps } from '@rneui/base/dist/Image/Image';

export { Image };
export type { ImageProps };

const ThemedImage = Object.assign(withTheme<ImageProps>(Image, 'Image'), {
  getSize: ImageNative.getSize,
  getSizeWithHeaders: ImageNative.getSizeWithHeaders,
  prefetch: ImageNative.prefetch,
  abortPrefetch: ImageNative.abortPrefetch,
  queryCache: ImageNative.queryCache,
  resolveAssetSource: ImageNative.resolveAssetSource,
});

export default ThemedImage;
