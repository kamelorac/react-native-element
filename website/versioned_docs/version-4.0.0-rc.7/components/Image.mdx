---
id: image
title: Image
---

import { IconsStyle } from "@site/src/components/Docs_Icons";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import CodeBlock from "@theme/CodeBlock";
import { BiInfoCircle } from "react-icons/bi";
import Usage from "../component_usage/Image.mdx";

<IconsStyle />

Drop-in replacement for the standard React Native Image component that displays
images with a placeholder and smooth image load transitioning.

## Usage

<div class="row inline-flex-center">
  <div class="col col--3">
    <h4>Import</h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`import { Image } from '@rneui/themed';`}</CodeBlock>
  </div>
  <div class="col col--3">
    <h4>
      Theme Key{" "}
      <a href="../customizing#using-themeprovider">
        <BiInfoCircle />
      </a>
    </h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`Image`}</CodeBlock>
  </div>
</div>

<Usage />

## Props

:::note
Includes all [React Native Image](https://reactnative.dev/docs/image#props) props.
:::

<div class='table-responsive'>

| Name                     | Type                                                        | Default                                           | Description                                           |
| ------------------------ | ----------------------------------------------------------- | ------------------------------------------------- | ----------------------------------------------------- |
| `Component`              | React Component                                             | `Press handlers present then Pressable else View` | Define the component passed to image.                 |
| `ImageComponent`         | typeof Component                                            |                                                   | Specify a different component as the Image component. |
| `PlaceholderContent`     | `ReactElement<any, string` \| `JSXElementConstructor<any>>` |                                                   | Content to load when Image is rendering.              |
| `childrenContainerStyle` | View Style                                                  | `null`                                            | Additional styling for the children container.        |
| `containerStyle`         | View Style                                                  |                                                   | Additional styling for the container.                 |
| `onLongPress`            | GestureResponderEventHandler                                |                                                   | Called when a long-tap gesture is detected.           |
| `onPress`                | GestureResponderEventHandler                                |                                                   | Called when a single tap gesture is detected.         |
| `onPressIn`              | GestureResponderEventHandler                                |                                                   | Called when a touch is engaged before `onPress`.      |
| `onPressOut`             | GestureResponderEventHandler                                |                                                   | Called when a touch is released before `onPress`.     |
| `placeholderStyle`       | View Style                                                  |                                                   | Additional styling for the placeholder container.     |
| `pressableProps`         | PressableProps except click handlers                        | `None`                                            |                                                       |
| `transition`             | boolean                                                     |                                                   | Perform fade transition on image load.                |
| `transitionDuration`     | number                                                      | `360`                                             | Perform fade transition on image load.                |

</div>
