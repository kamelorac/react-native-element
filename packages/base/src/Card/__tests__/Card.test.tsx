import React from 'react';
import { Card } from '..';
import { renderWithWrapper } from '../../../.ci/testHelper';
import { Image } from '../../Image';

describe('Card Component', () => {
  it('should match snapshot', () => {
    const component = renderWithWrapper(<Card />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should have Card title without image', () => {
    const { toJSON, queryByText } = renderWithWrapper(
      <Card>
        <Card.Title>Card Title</Card.Title>
      </Card>
    );
    expect(queryByText('Card Title')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('should have Card title with image', () => {
    const { wrapper, queryByText } = renderWithWrapper(
      <Card>
        <Card.Title>HELLO WORLD</Card.Title>
        <Card.Divider />
        <Card.Image
          source={{
            uri: 'https://s3.amazonaws.com/uifaces/faces/twitter/ladylexy/128.jpg',
          }}
        />
      </Card>
    );
    expect(queryByText('HELLO WORLD')).toBeTruthy();
    expect(wrapper.findByType(Image).props.source).toMatchObject({
      uri: 'https://s3.amazonaws.com/uifaces/faces/twitter/ladylexy/128.jpg',
    });
  });

  it('should have Card with featured title', () => {
    const { queryByText } = renderWithWrapper(
      <Card>
        <Card.FeaturedTitle>featured title</Card.FeaturedTitle>
        <Card.FeaturedSubtitle>featured sub title</Card.FeaturedSubtitle>
        <Card.Image
          source={{
            uri: 'https://s3.amazonaws.com/uifaces/faces/twitter/ladylexy/128.jpg',
          }}
        />
      </Card>
    );
    expect(queryByText('featured title')).toBeTruthy();
    expect(queryByText('featured sub title')).toBeTruthy();
  });
});
