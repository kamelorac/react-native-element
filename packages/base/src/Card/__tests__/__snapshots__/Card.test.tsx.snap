// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should have Card title without image 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "transparent",
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#43484d",
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 15,
            "textAlign": "center",
          }
        }
        testID="cardTitle"
      >
        Card Title
      </Text>
    </View>
  </View>
</View>
`;

exports[`Card Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "transparent",
        }
      }
    />
  </View>
</View>
`;
